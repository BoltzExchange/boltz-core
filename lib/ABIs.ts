import ERC20ABI from '../out/ERC20.sol/ERC20.json';
import ERC20SwapABI from '../out/ERC20Swap.sol/ERC20Swap.json';
import EtherSwapABI from '../out/EtherSwap.sol/EtherSwap.json';
import RouterABI from '../out/Router.sol/Router.json';

export const ContractABIs = {
  ERC20: ERC20ABI.abi,
  ERC20Swap: ERC20SwapABI.abi,
  EtherSwap: EtherSwapABI.abi,
  Router: RouterABI.abi,
} as const;
