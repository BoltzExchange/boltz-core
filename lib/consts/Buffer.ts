export const Nonce = Buffer.from('00', 'hex');
export const EmptyScript = Buffer.alloc(0);
export const PrefixUnconfidential = Buffer.from('01', 'hex');
