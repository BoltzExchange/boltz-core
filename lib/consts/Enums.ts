export enum OutputType {
  Bech32 = 0,
  Compatibility = 1,
  Legacy = 2,
  Taproot = 3,
}
