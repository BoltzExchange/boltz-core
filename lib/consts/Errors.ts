const Errors = {
  TAPROOT_NOT_SUPPORTED: 'taproot not supported',
};

export default Errors;
