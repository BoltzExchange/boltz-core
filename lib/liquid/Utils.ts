import { confidential, TxOutput } from 'liquidjs-lib';
import { confidentialLiquid } from './Init';

export const getOutputValue = (
  output: TxOutput & {
    blindingPrivKey?: Buffer;
  },
): number => {
  return output.blindingPrivKey
    ? Number(
        confidentialLiquid.unblindOutputWithKey(output, output.blindingPrivKey)
          .value,
      )
    : confidential.confidentialValueToSatoshi(output.value);
};
