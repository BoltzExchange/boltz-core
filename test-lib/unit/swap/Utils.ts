import Errors from '../../../lib/consts/Errors';
import { OutputType } from '../../../lib/consts/Enums';
import { p2wshOutput, p2shOutput, p2shP2wshOutput } from '../../../lib/swap/Scripts';

export const getScriptHashFunction = (type: OutputType): (scriptHex: Buffer) => Buffer => {
  switch (type) {
    case OutputType.Bech32: return p2wshOutput;
    case OutputType.Legacy: return p2shOutput;
    case OutputType.Compatibility: return p2shP2wshOutput;
    case OutputType.Taproot: throw Errors.TAPROOT_NOT_SUPPORTED;
  }
};
