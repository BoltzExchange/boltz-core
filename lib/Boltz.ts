import { targetFee } from './TargetFee';
import { OutputType } from './consts/Enums';
import Networks from './consts/Networks';
import * as Types from './consts/Types';
import {
  ClaimDetails,
  RefundDetails,
  ScriptElement,
  TransactionOutput,
} from './consts/Types';
import { init } from './init';
import Musig from './musig/Musig';
import { constructClaimTransaction } from './swap/Claim';
import { detectPreimage } from './swap/PreimageDetector';
import { constructRefundTransaction } from './swap/Refund';
import reverseSwapScript from './swap/ReverseSwapScript';
import reverseSwapTree, {
  extractClaimPublicKeyFromReverseSwapTree,
  extractRefundPublicKeyFromReverseSwapTree,
} from './swap/ReverseSwapTree';
import * as Scripts from './swap/Scripts';
import { detectSwap } from './swap/SwapDetector';
import swapScript from './swap/SwapScript';
import swapTree, {
  extractClaimPublicKeyFromSwapTree,
  extractRefundPublicKeyFromSwapTree,
} from './swap/SwapTree';
import { compareTrees } from './swap/SwapTreeCompare';
import * as SwapTreeSerializer from './swap/SwapTreeSerializer';
import * as SwapUtils from './swap/SwapUtils';
import * as TaprootUtils from './swap/TaprootUtils';

export {
  Musig,
  Types,
  Networks,
  OutputType,
  ClaimDetails,
  ScriptElement,
  RefundDetails,
  TransactionOutput,
  Scripts,
  SwapUtils,
  TaprootUtils,
  SwapTreeSerializer,
  init,
  swapTree,
  targetFee,
  swapScript,
  detectSwap,
  compareTrees,
  detectPreimage,
  reverseSwapTree,
  reverseSwapScript,
  constructClaimTransaction,
  constructRefundTransaction,
  extractClaimPublicKeyFromSwapTree,
  extractRefundPublicKeyFromSwapTree,
  extractClaimPublicKeyFromReverseSwapTree,
  extractRefundPublicKeyFromReverseSwapTree,
};
