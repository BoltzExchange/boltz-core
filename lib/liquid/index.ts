import { init } from './Init';
import Errors from './consts/Errors';
import Networks from './consts/Networks';
import { getOutputValue } from './Utils';
import { constructClaimTransaction } from './swap/Claim';
import { constructRefundTransaction } from './swap/Refund';

export {
  Errors,
  Networks,
  init,
  getOutputValue,
  constructClaimTransaction,
  constructRefundTransaction,
};
