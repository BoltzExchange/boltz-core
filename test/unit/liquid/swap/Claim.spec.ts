import zkp from '@vulpemventures/secp256k1-zkp';
import { getHexBuffer } from '../../../../lib/Utils';
import { OutputType } from '../../../../lib/consts/Enums';
import { LiquidClaimDetails } from '../../../../lib/liquid';
import { constructClaimTransaction, init } from '../../../../lib/liquid';
import { liquidClaimDetailsMap } from './ClaimDetails';

describe('Liquid Claim', () => {
  const testClaim = (utxos: LiquidClaimDetails[], fee: number) => {
    return constructClaimTransaction(
      utxos,
      getHexBuffer('00140000000000000000000000000000000000000000'),
      fee,
      false,
    );
  };

  beforeAll(async () => {
    init(await zkp());
  });

  test('should claim a P2WSH swap', () => {
    expect(
      testClaim([liquidClaimDetailsMap.get(OutputType.Bech32)!], 22).toHex(),
    ).toMatchSnapshot();
  });

  test('should claim multiple P2WSH swaps in one transaction', () => {
    const details = liquidClaimDetailsMap.get(OutputType.Bech32)!;

    expect(
      testClaim(
        [
          details,
          {
            ...details,
            vout: details.vout + 1,
          },
          {
            ...details,
            vout: details.vout + 2,
          },
        ],
        40,
      ).toHex(),
    ).toMatchSnapshot();
  });

  test.each`
    type                        | name
    ${OutputType.Compatibility} | ${'P2SH nested P2WSH'}
    ${OutputType.Legacy}        | ${'P2SH'}
  `(`should throw with $name inputs`, ({ type }) => {
    expect(() =>
      testClaim(
        [
          {
            type,
            redeemScript: Buffer.alloc(0),
          } as LiquidClaimDetails,
        ],
        1,
      ),
    ).toThrow('only Taproot or native SegWit inputs supported');
  });

  test('should throw with inconsistently blinded inputs', () => {
    expect(() =>
      testClaim(
        [
          {
            type: OutputType.Bech32,
            redeemScript: Buffer.alloc(0),
            blindingPrivateKey: Buffer.alloc(1),
          },
          {
            type: OutputType.Bech32,
            redeemScript: Buffer.alloc(0),
          },
        ] as LiquidClaimDetails[],
        1,
      ),
    ).toThrow('all or none inputs have to be blinded');
  });
});
