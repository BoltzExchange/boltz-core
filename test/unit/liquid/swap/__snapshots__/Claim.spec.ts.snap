// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Liquid Claim should claim a P2WSH swap 1`] = `"020000000101285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d7540000000000ffffffff02016d521c38ec1ea15734ae22b7c46064412829c0d0579f0a713d1c04ede979026f0100000000000007ba001600140000000000000000000000000000000000000000016d521c38ec1ea15734ae22b7c46064412829c0d0579f0a713d1c04ede979026f010000000000000016000000000000000003483045022100cacab778b7ed5c5ead44e080da85c11b63cbcb4ddbd025093fa5b68951b7fa90022063a4cbf19ea317c3a597c4f5aab86b7912fcc8814ba99984b38beb1176b078e30110b5b2dbb1f0663878ecbc20323b58b92c64a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac0000000000"`;

exports[`Liquid Claim should claim multiple P2WSH swaps in one transaction 1`] = `"020000000103285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d7540000000000ffffffff285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d7540100000000ffffffff285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d7540200000000ffffffff02016d521c38ec1ea15734ae22b7c46064412829c0d0579f0a713d1c04ede979026f010000000000001748001600140000000000000000000000000000000000000000016d521c38ec1ea15734ae22b7c46064412829c0d0579f0a713d1c04ede979026f0100000000000000280000000000000000034730440220417f50e9bdd2c6636cb35cc50c3899b6f34e425c2e8ac6ed158e789fd411cdd002207b3f2b570ebca0df06cd9af527d40fe411d9abc887da075a99f2344caaf136a70110b5b2dbb1f0663878ecbc20323b58b92c64a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac00000003483045022100985b2d0630eddbc0fdb80c9c7ca1a8cd0a4570d18011ad9b2bf5257963dda098022024db6ba6725e1bd98b723f85ac75a16b24e36f2808a24e534fc04da76ff961730110b5b2dbb1f0663878ecbc20323b58b92c64a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac00000003473044022012dc2e8b2ce674b623b0edc7f17aecd2cd654cd9d8dff6971a5820bb78ae816202200254e529ced3ed531439db18e58d35367431d203707b9e4481876620e76d0a250110b5b2dbb1f0663878ecbc20323b58b92c64a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac0000000000"`;
